tests/integrations/raftstore/test_hibernate.rs

```diff
@@ -410,3 +410,89 @@ fn test_hibernate_feature_gate() {
     // Leader can go to sleep as version requirement is met.
     assert!(!awakened.load(Ordering::SeqCst));
 }
+
+/// Tests when leader is demoted in a hibernated region, the region can recover automatically.
+#[test]
+fn test_leader_demoted_when_hibernated() {
+    let mut cluster = new_node_cluster(0, 4);
+    configure_for_hibernate(&mut cluster);
+    cluster.pd_client.disable_default_operator();
+    let r = cluster.run_conf_change();
+    cluster.pd_client.must_add_peer(r, new_peer(2, 2));
+    cluster.pd_client.must_add_peer(r, new_peer(3, 3));
+    cluster.must_put(b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(3), b"k1", b"v1");
+
+    cluster.must_transfer_leader(r, new_peer(1, 1));
+    // Demote a follower to learner.
+    cluster.pd_client.must_joint_confchange(
+        r,
+        vec![
+            (ConfChangeType::AddLearnerNode, new_learner_peer(3, 3)),
+            (ConfChangeType::AddLearnerNode, new_learner_peer(4, 4)),
+        ],
+    );
+    // So leader will not commit the demote request.
+    cluster.add_send_filter(CloneFilterFactory(
+        RegionPacketFilter::new(r, 1)
+            .msg_type(MessageType::MsgAppendResponse)
+            .direction(Direction::Recv),
+    ));
+
+    // So only peer 3 can become leader.
+    for id in 1..=2 {
+        cluster.add_send_filter(CloneFilterFactory(
+            RegionPacketFilter::new(r, id)
+                .msg_type(MessageType::MsgRequestPreVote)
+                .direction(Direction::Send),
+        ));
+    }
+    // Leave joint.
+    cluster.async_exit_joint(r).unwrap();
+    // Ensure peer 3 can campaign.
```

Comment:
Can we just transfer leadership to peer 3 then leave joint, while leaving the joint state, tikv will not check and reject leave joint if the leader will be demoted.

Reply:
> ...tikv will not check and reject leave joint if the leader will be demoted.

Then it should be a bug.

Reply:
https://github.com/tikv/tikv/pull/9533 add check to prevent it, PTAL